pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -430,6 +438,165 @@ func consumerRefMatches(consumer *corev1.ObjectReference, machine *machinev1.Mac
 	return true
 }
 
+// updateUserData fetches full ignition from MCS and updates the UserData secret reference
+func (a *Actuator) updateUserData(ctx context.Context, config *bmv1alpha1.BareMetalMachineProviderSpec,
+	machine *machinev1.Machine) error {
+
+	// Try to fetch the full ignition secret to see if it exists
+	machineUserDataSecret := &corev1.Secret{}
+	key := client.ObjectKey{
+		Name:      machine.Name + "-user-data",
+		Namespace: machine.Namespace,
+	}
+	// Secret already exists. Set secretReference and return
+	err := a.client.Get(ctx, key, machineUserDataSecret)
+	if err == nil {
+		// Set Machine's secretreference
+		config.UserData = &corev1.SecretReference{
+			Name:      machine.Name + "-user-data",
+			Namespace: machine.Namespace,
+		}
+		return nil
+	}
+
+	// userDataSecret is the pointer ignition originally
+	// created by openshift installer. Fetch it to extract
+	// its content (an URL containg full ignition and a
+	// CA cert for the URL's https connection)
+	userDataSecret := corev1.Secret{}
+	key = client.ObjectKey{
+		Name:      config.UserData.Name,
+		Namespace: config.UserData.Namespace,
+	}
+	err = a.client.Get(ctx, key, &userDataSecret)
+	if err != nil {
+		log.Printf("Cannot get %s secret: %s", config.UserData.Name, err.Error())
+		return err
+	}
+	encodedUserDataBytes := userDataSecret.Data["userData"]
+	var userData []byte
+	userData, err = base64.StdEncoding.DecodeString(string(encodedUserDataBytes))
+	if err != nil {
+		log.Printf("Error base64 decoding userData : %s", err.Error())
+		return fmt.Errorf("error base64 decoding userData: %w", err)
+	}
+
+	// Define an ignition data struct to unmarshal incoming json
+	type IgnitionData struct {
+		Ignition struct {
+			Config struct {
+				Append []struct {
+					Source string
+				}
+			}
+			Security struct {
+				TLS struct {
+					CertificateAuthorities []struct {
+						Source string
+					}
+				}
+			}
+		}
+	}
+	var ignData IgnitionData
+
+	if err := json.Unmarshal(userData, &ignData); err != nil {
+		log.Printf(string(userData))
+		log.Printf("Cannot unmarshal ignition config from userData : %s", err.Error())
```

Comment:
Something in the error messages produced here needs to include the name of the secret we were looking at and unable to parse. I don't think we need to log here at all, because returning an error should cause the caller to log.

Reply:
One alternative to parsing ignition ourselves might be to use the ignition code to render the config, which will resolve all the appends.

https://github.com/coreos/ignition/blob/master/internal/exec/engine.go#L268

This would probably be the ideal way to handle things, in case the installer ever changes the ignition format, we'll be able to parse anything they throw at us

Reply:
It's not easy to see if any of this is exported out of ignition, we might have to write something on their side to do this. It'd solve the complaints from the installer team if we were able to do that

Reply:
+1. Made the change. PTAL